File: PoketDex/app/src/main/java/pt/rfsfernandes/ui/fragments/details_pager_fragments/StatsInfoFragment.java
package pt.rfsfernandes.ui.fragments.details_pager_fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import pt.rfsfernandes.R;
import pt.rfsfernandes.databinding.FragmentStatsInfoBinding;
import pt.rfsfernandes.model.pokemon.Pokemon;
import pt.rfsfernandes.model.pokemon.stats.Stats;
import pt.rfsfernandes.viewmodels.MainViewModel;


public class StatsInfoFragment extends Fragment {

  private FragmentStatsInfoBinding binding;
  private MainViewModel mMainViewModel;


  public StatsInfoFragment() {
    // Required empty public constructor
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
                           Bundle savedInstanceState) {
    binding = FragmentStatsInfoBinding.inflate(inflater, container, false);
    View view = binding.getRoot();
    mMainViewModel = new ViewModelProvider(requireActivity()).get(MainViewModel.class);
    return view;
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    initViewModel();
  }

  /**
   * Initiates viewModel observers
   */
  private void initViewModel() {

    mMainViewModel.getPokemonMutableLiveData().observe(getViewLifecycleOwner(), pokemon -> {
      if (pokemon != null) {
        mMainViewModel.isLoading(true);
        setStatViews(pokemon);
        mMainViewModel.isLoading(false);
      }
    });


  }

  /**
   * Sets each pokemon stat
   *
   * @param pokemon selected pokemon
   */
  private void setStatViews(Pokemon pokemon) {
    List<Stats> statsList = pokemon.getStatsList();

    if (statsList.get(0) != null) {
      binding.textViewStatsHp.setText(String.valueOf(statsList.get(0).getBaseStat()));
    }

    if (statsList.get(1) != null) {
      binding.textViewStatsAttack.setText(String.valueOf(statsList.get(1).getBaseStat()));
    }

    if (statsList.get(2) != null) {
      binding.textViewStatsDefense.setText(String.valueOf(statsList.get(2).getBaseStat()));
    }

    if (statsList.get(3) != null) {
      binding.textViewStatsSpecialAttack.setText(String.valueOf(statsList.get(3).getBaseStat()));
    }

    if (statsList.get(4) != null) {
      binding.textViewStatsSpecialDefense.setText(String.valueOf(statsList.get(4).getBaseStat()));
    }

    if (statsList.get(5) != null) {
      binding.textViewStatsSpeed.setText(String.valueOf(statsList.get(5).getBaseStat()));
    }

    if(getContext() != null) {
      float weightInKg = pokemon.getWeight() / 10f;
      binding.textViewStatsWeight.setText(String.format(Locale.getDefault(), "%.1f%s", weightInKg,
          getContext().getString(R.string.kilograms)));

      float heightInMeters = pokemon.getHeight() / 10f;
      binding.textViewStatsHeight.setText(String.format(Locale.getDefault(), "%.1f%s", heightInMeters,
          getContext().getString(R.string.meters)));
    }

    binding.textViewStatsBaseExperience.setText(String.valueOf(pokemon.getBaseExperience()));

  }

}

File: PocketDex/app/src/main/java/pt/rfsfernandes/pocketdex/ui/fragments/details_pager_fragments/StatsInfoFragment.kt
package pt.rfsfernandes.pocketdex.ui.fragments.details_pager_fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import org.koin.androidx.viewmodel.ext.android.sharedViewModel
import pt.rfsfernandes.pocketdex.R
import pt.rfsfernandes.pocketdex.databinding.FragmentStatsInfoBinding
import pt.rfsfernandes.pocketdex.model.pokemon.Pokemon
import pt.rfsfernandes.pocketdex.model.pokemon.stats.Stats
import pt.rfsfernandes.pocketdex.viewmodels.MainViewModel
import java.util.*

class StatsInfoFragment : Fragment() {
    private lateinit var binding: FragmentStatsInfoBinding
    private val mMainViewModel: MainViewModel by sharedViewModel()
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentStatsInfoBinding.inflate(inflater, container, false)
        val view: View = binding.root
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViewModel()
    }

    /**
     * Initiates viewModel observers
     */
    private fun initViewModel() {
        mMainViewModel.pokemonMutableLiveData.observe(viewLifecycleOwner) { pokemon: Pokemon? ->
            if (pokemon != null) {
                mMainViewModel.isLoading(true)
                setStatViews(pokemon)
                mMainViewModel.isLoading(false)
            }
        }
    }

    /**
     * Sets each pokemon stat
     *
     * @param pokemon selected pokemon
     */
    private fun setStatViews(pokemon: Pokemon) {
        val statsList: List<Stats?> = pokemon.statsList
        if (statsList[0] != null) {
            binding.textViewStatsHp.text = statsList[0]!!.baseStat.toString()
        }
        if (statsList[1] != null) {
            binding.textViewStatsAttack.text = statsList[1]!!.baseStat.toString()
        }
        if (statsList[2] != null) {
            binding.textViewStatsDefense.text = statsList[2]!!.baseStat.toString()
        }
        if (statsList[3] != null) {
            binding.textViewStatsSpecialAttack.text = statsList[3]!!.baseStat.toString()
        }
        if (statsList[4] != null) {
            binding.textViewStatsSpecialDefense.text = statsList[4]!!.baseStat.toString()
        }
        if (statsList[5] != null) {
            binding.textViewStatsSpeed.text = statsList[5]!!.baseStat.toString()
        }
        if (context != null) {
            val weightInKg = pokemon.weight / 10f
            binding.textViewStatsWeight.text = String.format(Locale.getDefault(), "%.1f%s", weightInKg,
                    requireContext().getString(R.string.kilograms))
            val heightInMeters = pokemon.height / 10f
            binding.textViewStatsHeight.text = String.format(Locale.getDefault(), "%.1f%s", heightInMeters,
                    requireContext().getString(R.string.meters))
        }
        binding.textViewStatsBaseExperience.text = pokemon.baseExperience.toString()
    }
}